packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -565,7 +593,19 @@ - (CVPixelBufferRef)copyPixelBuffer {
     }
   }
 
-  return buffer;
+  // Calling textureFrameAvailable only from within displayLinkFired would require a non-trivial
+  // solution to minimize missed video frames due to race between displayLinkFired, copyPixelBuffer
+  // and place where is _textureFrameAvailable reset to false in the flutter engine. Ideally
+  // FlutterTexture would support mode of operation where the copyPixelBuffer is called always or
+  // some other alternative, instead of on demand by calling textureFrameAvailable.
+  if (self.displayLink.running) {
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [self.frameUpdater.registry textureFrameAvailable:self.frameUpdater.textureId];
```

Comment:
> Display link is still usable for obtaining actual `duration` and for starting after pause.

Neither of those things require a regularly firing display link; one is just a getter, and the other is a one-time call. This version appears to have two sets of regular calls: the display link callback, on a set cadence that we determine based on the video, and then this, which is as fast as possible no matter what the refresh rate, frame rate, or elapsed time are.

I understand the goals of the previous iterations of this PR, but I don't understand unconditionally driving buffer copies as fast as possible no matter what, or constantly telling the engine that there are new frames available regardless of whether or not that's true.

Reply:
> on a set cadence that we determine based on the video

It is actually based on display refresh rate.

> which is as fast as possible no matter what the refresh rate, frame rate, or elapsed time are

This too, it is not faster than display refresh rate, it is at display refresh rate.

Reply:
> It is actually based on display refresh rate.

Right, sorry; I was thinking misremembering the logic used to set `videoComposition.frameDuration` logic as feeding into the display link.

> This too, it is not faster than display refresh rate, it is at display refresh rate.

How is an unconditional and immediate call every time a frame is provided the same as the refresh rate?

Reply:
`copyPixelBuffer` is called as a reaction to display link callback (maybe not `CADisplayLink` maybe rather something like `CAMetalDisplayLink`). `textureFrameAvailable` just changes the flag for the engine to know that it should call `copyPixelBuffer` when rendering the next frame to obtain the newest pixel buffer. (If it were to call `copyPixelBuffer` immediately as a reaction to `textureFrameAvailable`, btw which would fix that race problem, then would it not be better to just have `textureFrameAvailable` to deliver that pixel buffer directly instead?)

Reply:
Also, thinking about this more, I don't understand how this call helps under the current engine implementation:
- the plugin is calling `textureFrameAvailable:` once per display link callback
- you're saying the engine calls `copyPixelBuffer` once per display link callback (assuming it's had a `textureFrameAvailable:` call since last time), via async dispatch to the main thread
  - thus this line would cause the plugin to call `textureFrameAvailable:`... at most once per display link callback, via async dispatch to the main thread, meaning we're now sometimes calling it twice per display link callback
- you said earlier that the engine consolidates `textureFrameAvailable:` calls

Given that these are async, it seems like this in practice (under the current engine implementation) just sprinkles random extra calls to `textureFrameAvailable:` with no particular timing guarantees as to where they would fall relative to the next display link call and `copyPixelBuffer` call. Am I missing something, or is that accurate and to the extent that this helps it's just from a "more calls is more calls" standpoint rather than any conceptual fix?

Reply:
> So what else could be used to obtain actual `duration` instead of display link?

In a model where the plugin is constantly immediately asking the engine to request a new frame I don't really understand why you would even need `duration` any more, or why we need the complicated heuristic about what time to use. It seems like if we are having the engine call back as fast as it can, then we would always want to show the current media time, because it seems like by definition if we're more than a frame off we have lost a frame anyway, so I don't see why wouldn't just jump to the current frame, making that logic much simpler. (Assuming we changed the engine contract such that this would be a safe implementation in the first place.)

What does the `duration`-based heuristic accomplish in this scenario?

Reply:
So let's assume there is implementation which calls `copyPixelBuffer` as soon as possible after `textureFrameAvailable`, so rate at which `copyPixelBuffer` will be called may be for example let's say 1000x per second (about 16x per screen refresh on 60 Hz display). This means that timestamps at which are video buffers "discovered" are just at most some 1ms off, which may be smaller spread (and less frame drops or at least less often) than using current time in implementation in this PR but much higher than using that duration (so-called) "heuristic". Now the higher the actual rate (lets say 10000x, 1000000x, but this is really uncontrollable) the smaller the spread and it could asymptotically approach what that "heuristic" achieves. But this would be highly inefficient and unnecessary.

Maybe I am not entirely sure what your point is with this.

> What does the duration-based heuristic accomplish in this scenario?

If such theoretical implementation is a concern I proposed frame rate limiting in the previous comment to resolve this. This would enforce that `copyPixelBuffer` will not be called 1000x per second but at most 60x per second on a 60 Hz display. And then we are back at start as we want duration for frame rate limiting and also for that "heuristic".

> In a model where the plugin is constantly immediately asking the engine to request a new frame I don't really understand why you would even need duration any more

The model in this PR is that it expects a `copyPixelBuffer` call at the next screen refresh. Again if the problem is that there is no such saying in contract I wrote about two possible ways to fix that in the previous comment (frame rate limiting or changing contract). Duration is needed to calculate next timestamp at which to sample because even if engine calls `copyPixelBuffer` as soon as possible after `textureFrameAvailable` it will still have jitter in currently observed time which will cause irregular sampling and frame drops (I suppose this part is clear?).

> or why we need the complicated heuristic about what time to use

It just does simple addition to maintain time where each next screen refresh it will be advanced by mostly fixed amount. Maybe that `if` block may look a little more complicated than it really is. It just says if our maintained time is too away from real time then just reset/adjust it accordingly.

> (Assuming we changed the engine contract such that this would be a safe implementation in the first place.)

You mean implementation of `FlutterTexture` in the engine or implementation of `copyPixelBuffer` on the plugin side?

Reply:
> So let's assume there is implementation which calls `copyPixelBuffer` as soon as possible after `textureFrameAvailable`, so rate at which `copyPixelBuffer` will be called may be for example let's say 1000x per second

That's not the case I'm asking about. I'm asking about the PR's current implementation (where every call to `copyPixelBuffer` immediately starts the process of calling `textureFrameAvailable` again) in conjunction with the assumptions you are currently making about the engine implementation (that `copyPixelBuffer` is being called at each screen refresh).

And under that model:

> it will still have jitter in currently observed time which will cause irregular sampling and frame drops (I suppose this part is clear?).

this is not clear to me, no. Is the issue that the actual execution time of `copyPixelBuffer` is actually far enough out of sync with the display link target time that it causes observable frame offsets? And so we're attempting to work around not having access to the target display time from the engine display link?

Reply:
> assumptions you are currently making about the engine implementation (that copyPixelBuffer is being called at each screen refresh)

Actually no such assumptions are really required for this PR to work fine. Maybe the only assumption is that `copyPixelBuffer` will not be called at too high a rate because screen refresh interval is enough and this is only due to avoiding inefficiency (like 1000 calls per second). But mentioned frame rate limiting should enforce that.

> this is not clear to me, no. Is the issue that the actual execution time of copyPixelBuffer is actually far enough out of sync with the display link target time that it causes observable frame offsets? And so we're attempting to work around not having access to the target display time from the engine display link?

By "frame offsets" you mean something as high as 1/60 or 2/60 of second (on 60 Hz display)? For example during the first screen refresh interval is `copyPixelBuffer` called twice then no call for two next refreshes then thrice and so on but on average there are still 60 calls per second? You mean something like this?

Reply:
No, I did not observe such high deviations. But much smaller deviations (than screen refresh period) are enough to give different frames by using `CACurrentMediaTime` vs something more regular. All calls to `CACurrentMediaTime` may return timestamp strictly between timestamps of current and next screen refresh but time points of frames in video are shifted randomly to this. Here is example of irregular sampling (like with `CACurrentMediaTime`) where video frame 2 was shown twice while frame 3 was dropped in contrast with regular sampling (like with `targetTime` in this PR):

```
screen refresh number:   -1---|---2---|---3---|---4---|---
irregular sampling:       v        v    v           v
video frame number:      |---1---|---2---|---3---|---4---|
regular sampling:          ^       ^       ^       ^
```

Reply:
Thanks, that's extremely helpful! I understand the issue much more clearly now. It would be a great idea to put that ASCII diagram into the comment in the code.

I think the next step is to file an issue to explore with the engine team whether making the `copyPixelBuffer`/`textureFrameAvailable` more explicit about call patterns (e.g., saying that calls will not be more frequent than ~the screen refresh rate) is something the engine team is comfortable with, so we know whether or not we need to build limiting logic at the plugin level.